h2o-algos/src/main/java/hex/word2vec/WordVectorTrainer.java

```diff
@@ -1,124 +1,80 @@
-      _HBWTCode = null;
-      _HBWTPoint = null;
-    } else { //HSM
-      _negExCnt = 0;
-      _unigramTable = null;
-      _HBWTCode = input._HBWTCode;
-      _HBWTPoint = input._HBWTPoint;
-    }
+    _initLearningRate = input.getParams()._initLearningRate;
+
+    _vocabWordCount = input._vocabWordCount;
+    _prevTotalProcessedWords = input._totalProcessedWords;
+
+    _syn0 = input._syn0;
+    _syn1 = input._syn1;
+    _curLearningRate = calcLearningRate(_initLearningRate, _epochs, _prevTotalProcessedWords, _vocabWordCount);
   }
-  final public Word2VecModelInfo getModelInfo() { return _output; }
 
   @Override
   protected void setupLocal() {
-    _syn0 = _input._syn0;  _syn1 = _input._syn1;
-    _output = _input; //faster, good enough in this case (since the input was freshly deserialized by the Weaver)
-    _input = null;
-    _rand = new Random();
+    _vocab = ((Vocabulary) DKV.getGet(_vocabKey))._data;
+    _wordCounts = ((WordCounts) DKV.getGet(_wordCountsKey))._data;
+    HBWTree t = DKV.getGet(_treeKey);
+    _HBWTCode = t._code;
+    _HBWTPoint = t._point;
     initExpTable();
-    buildVocabHashMap();
-    _curLearningRate = _output._curLearningRate;
-    _output.setLocallyProcessed(0);
-  }
-
-
-  private void buildVocabHashMap() {
-    Vec word = _vocab.vec(0);
-    _vocabHM = new NonBlockingHashMap<>((int)_vocab.numRows());
-    for(int i=0; i < _vocab.numRows(); i++) _vocabHM.put(word.atStr(new BufferedString(),i),i);
-  }
-
-  private void updateAlpha(int localWordCnt) {
-    _curLearningRate = _initLearningRate * (1 - (_output.getGloballyProcessed() + localWordCnt) / (float) (_epochs * _output._trainFrameSize + 1));
-    if (_curLearningRate < _initLearningRate * 0.0001F) _curLearningRate = _initLearningRate * 0.0001F;
-  }
-
-  /*
-   * All words in sentence should be in vocab
-   */
-  private int getSentence(int[] sentence, CStrChunk cs) {
-    Vec count = _vocab.vec(1);
-    BufferedString tmp = new BufferedString();
-    float ran;
-    int wIdx, sentIdx = 0;
-
-    int sentLen = (cs._len - 1 - _chkIdx);
-    if (sentLen >= MAX_SENTENCE_LEN) sentLen = MAX_SENTENCE_LEN;
-    else if (sentLen < MIN_SENTENCE_LEN) return 0;
-
-    for (; _chkIdx < cs._len; _chkIdx++) {
-      cs.atStr(tmp, _chkIdx);
-      if (!_vocabHM.containsKey(tmp)) continue; //not in vocab, skip
-      wIdx = _vocabHM.get(tmp);
-      if (_sentSampleRate > 0) {  // subsampling while creating a "_sentence"
-        // paper says: float ran = 1 - sqrt(sample / (vocab[word].cn / (float)trainWords));
-        ran = ((float) Math.sqrt(count.at8(wIdx) / (_sentSampleRate * _output._trainFrameSize)) + 1) * (_sentSampleRate * _output._trainFrameSize) / (float) count.at8(wIdx);
-        // paper says: ran > ....
-        if (ran < _rand.nextFloat()) continue;
-      }
-      sentence[sentIdx++] = wIdx;
-      if (sentIdx >= sentLen) break;
-    }
-
-    return sentLen;
+    _nodeProcessedWords = new IcedLong(0L);
   }
 
   // Precompute the exp() table
```

Comment:
Minor: can we store this table statically? It does not seem to depend on input parameters.